packages/xstate-react/src/useSelector.ts

```diff
@@ -0,0 +1,41 @@
+import { useEffect, useState } from 'react';
+import { ActorRef, Interpreter, Subscribable } from 'xstate';
+import { isActorWithState } from './useActor';
+import { getServiceSnapshot } from './useService';
+
+function isService(actor: any): actor is Interpreter<any, any, any, any> {
+  return 'state' in actor && 'machine' in actor;
+}
+
+const defaultCompare = (a, b) => a === b;
+
+export function useSelector<
+  TActor extends ActorRef<any, any>,
+  T,
+  TEmitted = TActor extends Subscribable<infer Emitted> ? Emitted : never
+>(
+  actor: TActor,
+  selector: (emitted: TEmitted) => T,
+  compare: (a: T, b: T) => boolean = defaultCompare,
+  getSnapshot: (a: TActor) => TEmitted = (a) =>
+    isService(a)
+      ? getServiceSnapshot(a)
+      : isActorWithState(a)
+      ? a.state
+      : undefined
+) {
+  const [selected, setSelected] = useState(() => selector(getSnapshot(actor)));
+
+  useEffect(() => {
+    const sub = actor.subscribe((emitted) => {
+      const nextSelected = selector(emitted);
+      if (!compare(selected, nextSelected)) {
+        setSelected(nextSelected);
+      }
+    });
+
+    return () => sub.unsubscribe();
+  }, [selector]);
```

Comment:
This dependency array suffers from 2 problems.

A minor thing is that this doesn't depend on `compare` and people can provide custom ones. In most real-life scenarios comparator should be pure, but it still might be worth putting it in this dependency array.

The second thing is much worse and defeats the main purpose of this PR (in non that unlikely scenarios) which is being able to optimize rerenders using `useSelector`. The created subscriber doesn't depend on `selected` and thus it suffers from the stale closure problem.

Reply:
Fixed this, I believe - please see a4e0c0a